packages/flutter/lib/src/widgets/app.dart

```diff
@@ -141,12 +141,15 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
     super.dispose();
   }
 
+  // On Android: the user has pressed the back button.
   @override
-  bool didPopRoute() {
+  Future<bool> didPopRoute() async {
     assert(mounted);
     NavigatorState navigator = _navigator.currentState;
     assert(navigator != null);
-    return navigator.pop();
+    if (!await navigator.willPop())
```

Comment:
Here's the way I understand didPopRoute's role in all of this.

If didPopRoute() returns true it just means that the pop request has been dealt with. In this case willPop() defeats the request, i.e. it's been dealt with, even if the route was disposed.